app/provider_details/provider_service.py

```diff
@@ -36,42 +42,77 @@ def init_app(
 
     @property
     def strategies(self):
+        """ This is a dictionary with notification types as the keys. """
         return self._strategies
 
     def validate_strategies(self) -> None:
         for notification_type, strategy in self.strategies.items():
             strategy.validate(notification_type)
 
     def get_provider(self, notification: Notification) -> ProviderDetails:
-        template_or_service_provider_id = self._get_template_or_service_provider_id(notification)
-        if template_or_service_provider_id:
-            provider = get_provider_details_by_id(template_or_service_provider_id)
+        """
+        Return an instance of ProviderDetails that is appropriate for the given notification.
+        """
 
-            if provider is None:
-                raise InvalidProviderException(f'provider {template_or_service_provider_id} could not be found')
-            elif not provider.active:
-                raise InvalidProviderException(f'provider {template_or_service_provider_id} is not active')
+        # This is a UUID (ProviderDetails primary key).
+        provider_id = self._get_template_or_service_provider_id(notification)
+        logger.debug("notification = %s", notification)
+        logger.debug("provider_id = %s", provider_id)
 
-        else:
-            provider_selection_strategy = self._strategies[NotificationType(notification.notification_type)]
-            provider = provider_selection_strategy.get_provider(notification)
+        if provider_id:
+            provider = get_provider_details_by_id(provider_id)
+        elif notification.notification_type != NotificationType.SMS:
+            # Use an alternative strategy to determine the provider.
+            provider_selection_strategy = self._strategies.get(NotificationType(notification.notification_type))
+            logger.debug("Provider selection strategy: %s", provider_selection_strategy)
+            provider = None if (provider_selection_strategy is None) \
+                else provider_selection_strategy.get_provider(notification)
 
-            if provider is None:
+            if provider is None and provider_selection_strategy is not None:
+                # This exception message is more detailed than the messages below.
                 raise InvalidProviderException(
-                    f'provider strategy {provider_selection_strategy.get_label()} could not find a suitable provider'
+                    f"Could not determine a provider using strategy {provider_selection_strategy.get_label()}."
                 )
+        else:
+            # Do not use any other criteria to determine the provider.  See issue 944.
```

Comment:
This would require someone to go look up what 944 is. I'd just throw in a note about how the US doesn't really support dynamic SMS providers. You can leave the bit about 944.

Reply:
done